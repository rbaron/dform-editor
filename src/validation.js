
export const validateSchema = (schemaText, keyExtractor, allowedKeys = null)  => {
  let schema;

  try {
    schema = JSON.parse(schemaText)
  } catch(e) {
    throw new Error(`Could not parse JSON: ${e}`)
  }

  const allowedKeysSet = allowedKeys ? new Set([...allowedKeys]) : null
  _traverseAndCheck(schema, keyExtractor, allowedKeysSet)

  return schema;
}

const _traverseAndCheck = (schema, keyExtractor, allowedKeysSet, seenKeys = new Set()) => {
  // Validate fields
  schema.fields.forEach(field => _validateField(field, keyExtractor))

  // Validate conds
  _validateConds(schema, seenKeys)

  const keys = schema.fields.map(keyExtractor)

  // Validate forbidden keys
  _validateKeys(keys, allowedKeysSet)

  const newSeenKeys = new Set([...seenKeys, ...keys])

  if (schema.children) {
    return schema.children.every(s => _traverseAndCheck(s, keyExtractor, allowedKeysSet, newSeenKeys))
  } else {
    return true
  }
}

const _validateField = (field, keyExtractor) => {
  if (!field.type) {
      throw new Error(`Field is missing type`)
  } else if (!keyExtractor(field)) {
      throw new Error(`Field is missing key`)
  }
}

const _validateConds = (schema, seenKeys) => {
  if (!schema.cond) {
    throw new Error(`Missing cond`)
  } else if (schema.cond.field_id && !seenKeys.has(schema.cond.field_id)) {
    throw new Error(`Input with key ${schema.cond.field_id} does not exist at this point`)
  } else if (schema.cond.conds) {
    schema.cond.conds.forEach(cond => {
      if (!seenKeys.has(cond.field_id)) {
        throw new Error(`Input with key ${cond.field_id} does not exist at this point`)
      }
    })
  }
}

const _validateKeys = (keys, allowedKeysSet) => {
  if (allowedKeysSet) {
    keys.forEach(k => {
      if (!allowedKeysSet.has(k)) {
        throw new Error(`Key ${k} is not allowed`)
      }
    })
  }
}
