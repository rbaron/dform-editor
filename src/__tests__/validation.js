import { validateSchema } from '../validation'

describe('validateSchema with invalid JSON', () => {
  it('raises if invalid JSON', () => {
    const schemaText = "{"
    expect(() => validateSchema(schemaText)).toThrowErrorMatchingSnapshot()
  });
})

describe('validateSchema with inputs not define previously', () => {
  it('raises if a cond references a non-existing input in compound conds', () => {
    const schemaText = JSON.stringify({
      cond: {
        type: 'always_true',
      },
      fields: [{
        type: 'boolean',
        id: 'input1',
        label: 'input1label',
      }],
      children: [{
        cond: {
          type: 'and',
          conds: [{
              type: 'truthy',
              field_id: 'input1',
          }, {
              type: 'truthy',
              // This input does not exist
              field_id: 'input2',
          }],
        },
        fields: [{
            type: 'boolean',
            id: 'input3',
            label: 'input3label',
        }],
      }],
    });

    const keyExtractor = item => item.id

    expect(() => validateSchema(schemaText, keyExtractor)).toThrowError(
      'Input with key input2 does not exist at this point')
  });

  it('raises if a cond references a non-existing input', () => {
    const schemaText = JSON.stringify({
      cond: {
        type: 'always_true',
      },
      fields: [{
        type: 'boolean',
        id: 'input1',
        label: 'input1label',
      }],
      children: [{
        cond: {
          type: 'truthy',
          // This input does not exist at this point
          field_id: 'input3',
        },
        fields: [{
            type: 'boolean',
            id: 'input3',
            label: 'input3label',
        }],
      }],
    });

    const keyExtractor = item => item.id

    expect(() => validateSchema(schemaText, keyExtractor)).toThrowError(
      'Input with key input3 does not exist at this point')
  });

  it('works with keyExtractor for label', () => {
    const schemaText = JSON.stringify({
      cond: {
        type: 'always_true',
      },
      fields: [{
        type: 'boolean',
        id: 'input1',
        label: 'input1label',
      }],
      children: [{
        cond: {
          type: 'truthy',
          field_id: 'input1label',
        },
        fields: [{
            type: 'boolean',
            id: 'input3',
            label: 'input3label',
        }],
      }],
    });

    const keyExtractor = item => item.label

    expect(validateSchema(schemaText, keyExtractor)).toBeTruthy();
  });

  it('works with keyExtractor for id', () => {
    const schemaText = JSON.stringify({
      cond: {
        type: 'always_true',
      },
      fields: [{
        type: 'boolean',
        id: 'input1',
        label: 'input1label',
      }],
      children: [{
        cond: {
          type: 'truthy',
          field_id: 'input1',
        },
        fields: [{
            type: 'boolean',
            id: 'input3',
            label: 'input3label',
        }],
      }],
    });

    const keyExtractor = item => item.id

    expect(validateSchema(schemaText, keyExtractor)).toBeTruthy();
  });
});

describe('validateSchema with allowedKeys', () => {
  it('raises if a forbidden key is used', () => {
    const schemaText = JSON.stringify({
      cond: {
        type: 'always_true',
      },
      fields: [{
        type: 'boolean',
        id: 'input1',
        label: 'input1label',
      }],
      children: [{
        cond: {
          type: 'and',
          conds: [{
              type: 'truthy',
              field_id: 'input1',
          }],
        },
        fields: [{
            type: 'boolean',
            id: 'input3',
            label: 'input3label',
        }],
      }],
    });

    const keyExtractor = item => item.id

    const allowedKeys = new Set(['input1'])

    expect(() => validateSchema(schemaText, keyExtractor, allowedKeys)).toThrowError(
      'Key input3 is not allowed')
  });

  it('raises if a forbidden key is used with keyExtractor for label', () => {
    const schemaText = JSON.stringify({
      cond: {
        type: 'always_true',
      },
      fields: [{
        type: 'boolean',
        label: 'input1label',
      }],
      children: [{
        cond: {
          type: 'and',
          conds: [{
              type: 'truthy',
              field_id: 'input1label',
          }],
        },
        fields: [{
            type: 'boolean',
            label: 'input3label',
        }],
      }],
    });

    const keyExtractor = item => item.label

    const allowedKeys = new Set(['input1label'])

    expect(() => validateSchema(schemaText, keyExtractor, allowedKeys)).toThrowError(
      'Key input3label is not allowed')
  });

  it('works', () => {
    const schemaText = JSON.stringify({
      cond: {
        type: 'always_true',
      },
      fields: [{
        type: 'boolean',
        id: 'input1',
        label: 'input1label',
      }],
      children: [{
        cond: {
          type: 'and',
          conds: [{
              type: 'truthy',
              field_id: 'input1',
          }],
        },
        fields: [{
            type: 'boolean',
            id: 'input3',
            label: 'input3label',
        }],
      }],
    });

    const keyExtractor = item => item.id

    const allowedKeys = new Set(['input1', 'input3'])

    expect(validateSchema(schemaText, keyExtractor, allowedKeys)).toBeTruthy();
  });

  it('works with keyExtractor for label', () => {
    const schemaText = JSON.stringify({
      cond: {
        type: 'always_true',
      },
      fields: [{
        type: 'boolean',
        label: 'input1label',
      }],
      children: [{
        cond: {
          type: 'and',
          conds: [{
              type: 'truthy',
              field_id: 'input1label',
          }],
        },
        fields: [{
            type: 'boolean',
            label: 'input3label',
        }],
      }],
    });

    const keyExtractor = item => item.label

    const allowedKeys = new Set(['input1label', 'input3label'])

    expect(validateSchema(schemaText, keyExtractor, allowedKeys)).toBeTruthy();
  });
});

describe('validateSchema with invalid fields', () => {
  it('raises if a child field is missing type', () => {
    const schemaText = JSON.stringify({
      cond: {
        type: 'always_true',
      },
      fields: [{
        type: 'boolean',
        id: 'input1',
        label: 'input1label',
      }],
      children: [{
        cond: {
          type: 'and',
          conds: [{
              type: 'truthy',
              field_id: 'input1',
          }],
        },
        fields: [{
            // Missing type
            //type: 'boolean',
            id: 'input3',
            label: 'input3label',
        }],
      }],
    });

    const keyExtractor = item => item.id

    expect(() => validateSchema(schemaText, keyExtractor, null)).toThrowError(
      'Field is missing type')
  });

  it('raises if a child field is missing key', () => {
    const schemaText = JSON.stringify({
      cond: {
        type: 'always_true',
      },
      fields: [{
        type: 'boolean',
        label: 'input1label',
      }],
      children: [{
        cond: {
          type: 'and',
          conds: [{
              type: 'truthy',
              field_id: 'input1label',
          }],
        },
        fields: [{
            type: 'boolean',
            // Missing key
            //label: 'input3label',
        }],
      }],
    });

    const keyExtractor = item => item.label

    expect(() => validateSchema(schemaText, keyExtractor, null)).toThrowError(
      'Field is missing key')
  });

  it('raises if one schema field is missing key', () => {
    const schemaText = JSON.stringify({
      cond: {
        type: 'always_true',
      },
      fields: [{
        type: 'boolean',
        // Missing key
        //label: 'input1label',
      }],
    });

    const keyExtractor = item => item.label

    expect(() => validateSchema(schemaText, keyExtractor, null)).toThrowError(
      'Field is missing key')
  });
});

describe('validateSchema with invalid conds', () => {
  it.only('raises if missing cond', () => {
    const schemaText = JSON.stringify({
      // Missing cond
      //cond: {
      //  type: 'always_true',
      //},
      fields: [{
        type: 'boolean',
        id: 'input1',
        label: 'input1label',
      }],
    });

    const keyExtractor = item => item.id

    expect(() => validateSchema(schemaText, keyExtractor, null)).toThrowError(
      'Missing cond')
  });
});
