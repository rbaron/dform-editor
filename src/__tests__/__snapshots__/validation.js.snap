// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateSchema with invalid JSON raises if invalid JSON 1`] = `"Could not parse JSON: SyntaxError: Unexpected end of JSON input"`;
